packages/@react-spectrum/button/src/ToggleButton.tsx

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2020 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {classNames, SlotProvider, useFocusableRef, useStyleProps} from '@react-spectrum/utils';
+import {FocusableRef} from '@react-types/shared';
+import {FocusRing} from '@react-aria/focus';
+import {mergeProps} from '@react-aria/utils';
+import React from 'react';
+import {SpectrumToggleButtonProps} from '@react-types/button';
+import styles from '@adobe/spectrum-css-temp/components/button/vars.css';
+import {Text} from '@react-spectrum/text';
+import {useButton, useToggleButton} from '@react-aria/button';
+import {useHover} from '@react-aria/interactions';
+import {useProviderProps} from '@react-spectrum/provider';
+import { useToggleState } from '@react-stately/toggle';
+
+function ToggleButton(props: SpectrumToggleButtonProps, ref: FocusableRef<HTMLButtonElement>) {
+  props = useProviderProps(props);
+  let {
+    isQuiet,
+    isDisabled,
+    isEmphasized,
+    children,
+    autoFocus,
+    ...otherProps
+  } = props;
+
+  let domRef = useFocusableRef(ref);
+  let {buttonProps, isPressed} = useToggleButton(props, domRef);
+  let {hoverProps, isHovered} = useHover({isDisabled});
+  let {styleProps} = useStyleProps(otherProps);
+  let isTextOnly = React.Children.toArray(props.children).every(c => !React.isValidElement(c));
+  let state = useToggleState(props);
+
+  return (
+    <FocusRing focusRingClass={classNames(styles, 'focus-ring')} autoFocus={autoFocus}>
```

Comment:
I wonder if we should extract the common rendering stuff here into a base component that can be shared between ActionButton and ToggleButton?

Reply:
Guess it depends if Spectrum plans to keep them pretty similar or not. I'd probably lean towards keeping them separate, it's not that much code overlap

Reply:
Yeah true, either way.